src/Compilers/CSharp/Portable/BoundTree/LengthBasedStringSwitchData.cs

```diff
@@ -0,0 +1,313 @@
+        {
+            public readonly LabelSymbol label;
+            public readonly int selectedCharPosition;
+            public readonly ImmutableArray<(ConstantValue value, LabelSymbol label)> charCaseLabels;
+
+            internal CharJumpTable(LabelSymbol label, int selectedCharPosition, ImmutableArray<(ConstantValue value, LabelSymbol label)> charCaseLabels)
+            {
+                Debug.Assert(charCaseLabels.All(c => c.value.IsChar) && charCaseLabels.Length > 0);
+
+                this.label = label;
+                this.selectedCharPosition = selectedCharPosition;
+                this.charCaseLabels = charCaseLabels;
+            }
+        }
+
+        internal struct StringJumpTable
+        {
+            public readonly LabelSymbol label;
+            public readonly ImmutableArray<(ConstantValue value, LabelSymbol label)> stringCaseLabels;
+
+            internal StringJumpTable(LabelSymbol label, ImmutableArray<(ConstantValue value, LabelSymbol label)> stringCaseLabels)
+            {
+                Debug.Assert(stringCaseLabels.All(c => c.value.IsString) && stringCaseLabels.Length > 0);
+
+                this.label = label;
+                this.stringCaseLabels = stringCaseLabels;
+            }
+        }
+
+        // Based on benchmarks, the previous hashcode-based approach arguably performs better
+        // when buckets have 6 candidates or more.
+        internal bool ShouldGenerateLengthBasedSwitch(int labelsCount)
+        {
+            return SwitchStringJumpTableEmitter.ShouldGenerateHashTableSwitch(labelsCount) &&
+                _stringJumpTables.All(t => t.stringCaseLabels!.Length <= 5);
+        }
+
+        internal static LengthBasedStringSwitchData Create(ImmutableArray<(ConstantValue value, LabelSymbol label)> inputCases)
+        {
+            Debug.Assert(inputCases.All(c => c.value.IsString && c.label is not null));
+
+            LabelSymbol? nullCaseLabel = null;
+            foreach (var inputCase in inputCases)
+            {
+                if (inputCase.value.IsNull)
+                {
+                    Debug.Assert(nullCaseLabel is null, "At most one null case per string dispatch");
+                    nullCaseLabel = inputCase.label;
+                }
+            }
+
+            var lengthCaseLabels = ArrayBuilder<(ConstantValue value, LabelSymbol label)>.GetInstance();
+            var charJumpTables = ArrayBuilder<CharJumpTable>.GetInstance();
+            var stringJumpTables = ArrayBuilder<StringJumpTable>.GetInstance();
+            foreach (var group in inputCases.Where(c => !c.value.IsNull).GroupBy(c => c.value.StringValue!.Length))
+            {
+                int stringLength = group.Key;
+                var labelForLength = CreateAndRegisterCharJumpTables(stringLength, group.Select(e => (e.value, e.label)).ToImmutableArray(), charJumpTables, stringJumpTables);
+                lengthCaseLabels.Add((ConstantValue.Create(stringLength), labelForLength));
```

Comment:
>(ConstantValue.Create(stringLength) [](http://example.com/codeflow?start=37&length=35)

Would it simplify anything if the values in `lengthCaseLabels` were `int` rather than `ConstantValue`? It looks like we only use the `ConstantValue` when we call into `EmitIntegerSwitchJumpTable` and at that point, we already converting from `(ConstantValue, LabelSymbol)` to `KeyValuePair<ConstantValue, LabelSymbol>`.